p2p/src/lib.rs

```diff
@@ -11,60 +11,80 @@ pub use network::{ConnectPeer, DisconnectPeer, NetworkBase, Post};
 use parity_scale_codec::{Decode, Encode};
 use thiserror::Error;
 
-/// Network is a main p2p start point.
+pub mod handshake;
 pub mod network;
-/// Peer is an endpoint to another node.
 pub mod peer;
 
 /// The main type to use for secure communication.
 pub type Network<T> = NetworkBase<T, X25519Sha256, ChaCha20Poly1305>;
 
 /// Errors used in [`crate`].
-#[derive(Clone, Debug, iroha_macro::FromVariant, Error, iroha_actor::Message)]
+#[derive(Debug, Error, iroha_actor::Message)]
 pub enum Error {
     /// Failed to read or write
     #[error("Failed IO operation.")]
     Io(#[source] std::sync::Arc<io::Error>),
```

Comment:
out of scope here, but it feels weird to return `Arc` in an error

Reply:
This is a bit of a hacky workaround for bad API decisions in `std`. `io::Error` holds a `dyn Error` object, meaning that you can't clone it (even though in theory you should be able to). Given that we need to store the error that caused the `Peer<T, K, E>` to enter the broken state, and raise an error up a call stack, this was the least painful way to handle the situation. 

The real solution is to convert an `std::io::Error` into a `crate::IoError` enum. As an added bonus, we would have more specific error messages, with instructions on how to fix them.